src/smt/theory_str_noodler/util.cc

```diff
@@ -150,4 +150,116 @@ namespace smt::noodler::util {
+        STRACE("str-split-word-to-automata",
+            tout << "split_word_to_automata with word:\n" << word << "\n";
+            tout << "and " << automata.size() << " automata\n";
+            if (is_trace_enabled("str-nfa")) {
+                for (auto aut : automata) {
+                    tout << *aut << "\n";
+                }
+            }
+        );
+
+        const unsigned NUM_OF_AUTOMATA = automata.size();
+        const unsigned LENGTH_OF_WORD = word.length();
+
+        unsigned current_automaton = 0; // index in automata, of an automaton whose word we are now computing
+        unsigned index_in_word = 0; // where in the word we are
+        std::vector<unsigned> backtracking_indexes; // vector that remembers to which index in word to backtrack to
+        mata::nfa::StateSet current_states{automata[0]->initial}; // the set of states where we are now in the current automaton
+        std::vector<mata::nfa::StateSet> backtracking_state_sets; // vector that remembers to which set of states to backtrack to
+        zstring current_word; // word we are currently building for current automaton
+        bool is_backtracked = false; // whether we just backtracked
+        
+        while (current_automaton != NUM_OF_AUTOMATA || index_in_word != LENGTH_OF_WORD) {
+            STRACE("str-split-word-to-automata", tout << "Current automaton and index in word: " << current_automaton << " " << index_in_word << "\n";);
+
+            // if we did not backtrack, we need to first check whether we can currently accept (if we are backtracking, we need to instead get longer word for the current automaton)
+            // also if the current automaton is the last automaton, we need to finish reading the word, so in that case, we only check after we read the whole word
+            if (!is_backtracked && (current_automaton != NUM_OF_AUTOMATA-1 || index_in_word == LENGTH_OF_WORD) && automata[current_automaton]->final.intersects_with(current_states)) {
+                // if we can accept, we save the current index, states, and word and move to the next automaton
+                STRACE("str-split-word-to-automata", tout << "Moving to next automaton\n";);
+                backtracking_indexes.push_back(index_in_word);
+                backtracking_state_sets.push_back(current_states);
+                words.push_back(current_word);
+                STRACE("str-split-word-to-automata",
+                    tout << "Current words:";
+                    for (const auto& word : words) {
+                        tout << " " << word;
+                    }
+                    tout << "\n";
+                );
+
+                ++current_automaton;
+                if (current_automaton != NUM_OF_AUTOMATA) {
+                    // index_in_word is not updated, as we stay at the same position
+                    current_states = mata::nfa::StateSet(automata[current_automaton]->initial);
```

Comment:
Every final state is connected to every initial state of the following automaton.

Reply:
Done